src/main/java/ui/components/pickers/LabelPickerState.java

```diff
@@ -0,0 +1,323 @@
+package ui.components.pickers;
+
+import backend.resource.TurboLabel;
+import util.Utility;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used to represent the state of the label picker. In addition to representing a state,
+ * it contains the logic which handles toggling/highlighting of labels and simplifies the retrieval of
+ * the resulting status for UI to display.
+ */
+public class LabelPickerState {
+    Set<String> initialLabels;
+    List<String> addedLabels;
+    List<String> removedLabels;
+    List<String> matchedLabels;
+    List<String> repoLabels;
+    OptionalInt currentSuggestionIndex;
+
+    public LabelPickerState(Set<String> initialLabels, List<String> repoLabels) {
+        this(initialLabels, new ArrayList<>(), new ArrayList<>(), new ArrayList<>(), repoLabels,
+                OptionalInt.empty());
+    }
+
+    private LabelPickerState(Set<String> initialLabels, List<String> addedLabels, List<String> removedLabels,
+                             List<String> matchedLabels, List<String> repoLabels) {
+        this(initialLabels, addedLabels, removedLabels, matchedLabels, repoLabels, OptionalInt.empty());
+    }
+
+    private LabelPickerState(Set<String> initialLabels, List<String> addedLabels, List<String> removedLabels,
+                             List<String> matchedLabels, List<String> repoLabels, OptionalInt currentSuggestionIndex) {
+        this.initialLabels = initialLabels;
+        this.addedLabels = addedLabels;
+        this.removedLabels = removedLabels;
+        this.matchedLabels = matchedLabels;
+        this.repoLabels = repoLabels;
+        this.currentSuggestionIndex = currentSuggestionIndex;
+    }
+
+    /**
+     * Determines resulting state based on given initial state and user input
+     *
+     * @param initialState the state of the label picker before the user has typed anything
```

Comment:
Why do you need initial state as a parameter? The constructor gives the initial state right?